### DISCUSSION TITLE
Reading offline maps from MBTiles

### DISCUSSION THREAD

**Discussion by DenisCarriere** [digidem/mapeo-desktop#550] [Ideas]
Just wanted to say this is an amazing initiative! Keep up the great work! 👍 

Was wondering if reading MBTiles instead of a Tile Servers would be beneficial for your application.

I've created a NodeJS [MBTiles binding](https://github.com/deniscarriere/mbtiles-offline) which I'm currently using in an Electron app, I'm using Leaflet and I've written a custom wrapper to Leaflet to read map tiles directly from the file system instead of reading them from a HTTP connection (local tile server).

Library which you will need:
- [`mbtiles-offline`](https://github.com/deniscarriere/mbtiles-offline) (tons of documentation to get you started)

If you need any help implementing this, let me know!

## Snippet to extend Leaflet to read MBTiles

Extending the `createTile` method in the Leaflet TileLayer Class 

```js
this.mbtiles.findOne([coords.x, coords.y, coords.z]).then(image => {
  if (!image) tile.src = L.Util.emptyImageUrl
  tile.src = window.URL.createObjectURL(new Blob([new Uint8Array(image)], {type: 'image/png'}))
  L.DomEvent.on(tile, 'load', Util.bind(this._tileOnLoad, this, done, tile))
  L.DomEvent.on(tile, 'error', Util.bind(this._tileOnError, this, done, tile))
})
```

**Comment by okdistribute** [comment]
Hi thanks for opening this @DenisCarriere!

Just to note, we are looking at this in the upcoming development cycle (~winter)

If you're still available to help implement this we'd be open to a pull request.
